src/main/java/de/tum/in/www1/artemis/service/LearningPathService.java

```diff
@@ -0,0 +1,411 @@
+        }
+        final var targetId = relation.getTailCompetency().getId();
+        String targetNodeId;
+        if (competencyToMatchCluster.containsKey(targetId)) {
+            targetNodeId = getMatchingClusterStartNodeId(competencyToMatchCluster.get(targetId));
+        }
+        else {
+            targetNodeId = getCompetencyStartNodeId(targetId);
+        }
+        final String relationEdgeId = getRelationEdgeId(sourceNodeId, targetNodeId);
+        // skip if relation has already been created (possible for edges linked to matching cluster start/end nodes)
+        if (!createdRelations.contains(relationEdgeId)) {
+            final var edge = new NgxLearningPathDTO.Edge(relationEdgeId, sourceNodeId, targetNodeId);
+            edges.add(edge);
+            createdRelations.add(relationEdgeId);
+        }
+    }
+
+    public static String getCompetencyStartNodeId(long competencyId) {
+        return "node-" + competencyId + "-start";
+    }
+
+    public static String getCompetencyEndNodeId(long competencyId) {
+        return "node-" + competencyId + "-end";
+    }
+
+    public static String getLectureUnitNodeId(long competencyId, long lectureUnitId) {
+        return "node-" + competencyId + "-lu-" + lectureUnitId;
+    }
+
+    public static String getExerciseNodeId(long competencyId, long exerciseId) {
+        return "node-" + competencyId + "-ex-" + exerciseId;
+    }
+
+    public static String getMatchingClusterStartNodeId(long matchingClusterId) {
+        return "matching-" + matchingClusterId + "-start";
+    }
+
+    public static String getMatchingClusterEndNodeId(long matchingClusterId) {
+        return "matching-" + matchingClusterId + "-end";
+    }
+
+    public static String getLectureUnitInEdgeId(long competencyId, long lectureUnitId) {
+        return "edge-" + competencyId + "-lu-" + getInEdgeId(lectureUnitId);
+    }
+
+    public static String getLectureUnitOutEdgeId(long competencyId, long lectureUnitId) {
+        return "edge-" + competencyId + "-lu-" + getOutEdgeId(lectureUnitId);
+    }
+
+    public static String getExerciseInEdgeId(long competencyId, long exercise) {
+        return "edge-" + competencyId + "-ex-" + getInEdgeId(exercise);
+    }
+
+    public static String getExerciseOutEdgeId(long competencyId, long exercise) {
+        return "edge-" + competencyId + "-ex-" + getOutEdgeId(exercise);
+    }
+
+    public static String getInEdgeId(long id) {
+        return "edge-" + id + "-in";
+    }
+
+    public static String getOutEdgeId(long id) {
+        return "edge-" + id + "-out";
+    }
+
+    public static String getRelationEdgeId(String sourceNodeId, String targetNodeId) {
+        return "edge-relation-" + sourceNodeId + "-" + targetNodeId;
+    }
+
+    public static String getDirectEdgeId(long competencyId) {
+        return "edge-" + competencyId + "-direct";
+    }
+
+    public LearningObject getRecommendation(LearningPath learningPath) {
```

Comment:
Looks good for a first draft but two ideas:
- This will always return the same object for a student. Is it better when a random object is returned, so there are a few options what to do next?
- I'm not too sure about the performance with 2000 students and multiple exercises and lecture units in a course. Maybe you have to move this method (partially) into a query instead e.g. of loading all students for each lecture unit.